ld-script-irom.patch

```diff
@@ -0,0 +1,116 @@
+     KEEP (*(SORT(.ctors.*)))
+     KEEP (*(.ctors))
++    __init_array_end = ABSOLUTE(.);
+     KEEP (*crtbegin.o(.dtors))
+     KEEP (*(EXCLUDE_FILE (*crtend.o) .dtors))
+     KEEP (*(SORT(.dtors.*)))
+@@ -151,6 +154,15 @@
+   } >dram0_0_seg :dram0_0_bss_phdr
+ /* __stack = 0x3ffc8000; */
+ 
++  .irom0.text : ALIGN(4)
++  {
++    _irom0_text_start = ABSOLUTE(.);
++    *(.irom0.literal .irom.literal .irom.text.literal .irom0.text .irom.text)
++    *libm.a:(.literal .text .literal.* .text.*)
++    *irom.a:(.literal .text .literal.* .text.*)
++    _irom0_text_end = ABSOLUTE(.);
++  } >irom0_0_seg :irom0_0_phdr
++
+   .text : ALIGN(4)
+   {
+     _stext = .;
+@@ -182,7 +194,7 @@
+     *(.entry.text)
+     *(.init.literal)
+     *(.init)
+-    *(.literal .text .literal.* .text.* .stub .gnu.warning .gnu.linkonce.literal.* .gnu.linkonce.t.*.literal .gnu.linkonce.t.*)
++    *(.literal .text .stub .gnu.warning .gnu.linkonce.literal.* .gnu.linkonce.t.*.literal .gnu.linkonce.t.*)
+     *(.fini.literal)
+     *(.fini)
+     *(.gnu.version)
+@@ -198,14 +210,8 @@
+     *(.gnu.linkonce.lit4.*)
+     _lit4_end = ABSOLUTE(.);
+   } >iram1_0_seg :iram1_0_phdr
+-
+-  .irom0.text : ALIGN(4)
+-  {
+-    _irom0_text_start = ABSOLUTE(.);
+-    *(.irom0.literal .irom.literal .irom.text.literal .irom0.text .irom.text)
+-    _irom0_text_end = ABSOLUTE(.);
+-  } >irom0_0_seg :irom0_0_phdr
+ }
+ 
+ /* get ROM code address */
+ INCLUDE "../ld/eagle.rom.addr.v6.ld"
++
+--- esp_iot_sdk_v1.5.2/ld/eagle.app.v6.new.2048.irom.ld.orig	2016-04-01 09:11:42.175496836 +0300
++++ esp_iot_sdk_v1.5.2/ld/eagle.app.v6.new.2048.irom.ld	2016-04-07 09:53:01.086596929 +0300
+@@ -102,11 +102,14 @@
+     *(.gnu.linkonce.e.*)
+     *(.gnu.version_r)
+     *(.eh_frame)
++    . = (. + 3) & ~ 3;
+     /*  C++ constructor and destructor tables, properly ordered:  */
++    __init_array_start = ABSOLUTE(.);
+     KEEP (*crtbegin.o(.ctors))
+     KEEP (*(EXCLUDE_FILE (*crtend.o) .ctors))
+     KEEP (*(SORT(.ctors.*)))
+     KEEP (*(.ctors))
++    __init_array_end = ABSOLUTE(.);
+     KEEP (*crtbegin.o(.dtors))
+     KEEP (*(EXCLUDE_FILE (*crtend.o) .dtors))
+     KEEP (*(SORT(.dtors.*)))
+@@ -151,6 +154,15 @@
+   } >dram0_0_seg :dram0_0_bss_phdr
+ /* __stack = 0x3ffc8000; */
+ 
++  .irom0.text : ALIGN(4)
++  {
++    _irom0_text_start = ABSOLUTE(.);
++    *(.irom0.literal .irom.literal .irom.text.literal .irom0.text .irom.text)
++    *libm.a:(.literal .text .literal.* .text.*)
++    *irom.a:(.literal .text .literal.* .text.*)
++    _irom0_text_end = ABSOLUTE(.);
++  } >irom0_0_seg :irom0_0_phdr
++
+   .text : ALIGN(4)
+   {
+     _stext = .;
+@@ -182,7 +194,7 @@
+     *(.entry.text)
+     *(.init.literal)
+     *(.init)
+-    *(.literal .text .literal.* .text.* .stub .gnu.warning .gnu.linkonce.literal.* .gnu.linkonce.t.*.literal .gnu.linkonce.t.*)
++    *(.literal .text .stub .gnu.warning .gnu.linkonce.literal.* .gnu.linkonce.t.*.literal .gnu.linkonce.t.*)
```

Comment:
Funny enough. I tested micropython with this. Since these sections are not in .text they end up in .lit. So actually their location in RAM is changed, but because they are still in RAM the linker script creates working binaries.